### DISCUSSION TITLE
Add extension to add block to block list model

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#14120] [Features and ideas]
Based on Umbraco v10.4.0 I had a use-case with a block list used in a sidebar and further a global block list (single block mode), where I wanted this the be added to block list model on specific page types.

I could add in the block list on these page type, but it would be tedious to copy/paste this to each block list and since this block always needed same content, I wanted to add this globally.

I found this works:

```
@{
	var blockItems = sidebar?.ToList();

	@if (settings?.CatalogSection is not null)
	{
		blockItems?.Insert(0, settings.CatalogSection);
	}

	BlockListModel? blockList = blockItems != null
		? new BlockListModel(blockItems)
		: null;

	@await Html.GetBlockListHtmlAsync(blockList, "sidebar")
}
```

but maybe we should have some additional extension methods on `BlockListModel`? eg. `Add()`, `Prepend()`, `Append()`, `Insert()`, `Remove()`. The ctor takes `IList<BlockListItem>`

![image](https://user-images.githubusercontent.com/2919859/232999386-a8427875-51ab-4a46-84b3-839605b18c96.png)

**Comment by abjerner** [comment]
The `BlockListModel` is currently immutable, which I think may be a good thing. Let's say you were able to make changes to it - such as adding new items - it wouldn't just be the model for the current request that you're modifying, but the cached block list value shared across multiple request.

Allowing developers to modify the model could therefore lead to unexpected results. Eg. with your example, if `BlockListModel` supporting adding new items to it, you'd have to check whether you previously added your custom block item. Otherwise you end up adding the same item one time for each request, which would be pretty bad.

But adding some constructor overloads would be a good idea. For instance a constructor that takes an `IEnumerable<BlockListItem>`. Currently you'd have to call `.ToList()` on it first. A new constructor overload would probably do that anyway internally, but then you don't have to do it in your code making it slightly simpler to work with.

**Comment by bjarnef** [comment]
I added an extension method:

```
public static class BlockListExtensions
{
	public static BlockListModel? InsertBlock(this BlockListModel? model, int index, BlockListItem item)
	{
		var items = model?.ToList() ?? new();

		items.Insert(index, item);

		BlockListModel? blockList = items.Count > 0
				? new BlockListModel(items)
				: null;

		return blockList;
	}
}
```

And then in view I have this:

```
@if (settings?.CatalogSection is not null)
{
    sidebar = sidebar.InsertBlock(0, settings.CatalogSection);
}

@await Html.GetBlockListHtmlAsync(sidebar, "sidebar")
```

**Comment by bergmania** [comment]
How common a use case is this? Not sure such extensions belong in the CMS, if it is a rare use case

**Reply by bjarnef** [reply]
Not sure, let's see how many votes this get 😁
However with single block mode, I think there may be a greater chance this is needed, as one may separate blocks into separate properties (block list - single mode) and combine this using `GetBlockListHtmlAsync()` + block list template.

So instead of a block list with three blocks, three properties with a single block.
This may be simpler if these blocks also are used in other contexts. E.g. single block could be useful as block to set open graph data a page as alternative to a composition and since the block list allow custom view.
